tests_config/run_spec.js

```diff
@@ -53,15 +54,16 @@ function run_spec(dirname, parsers, options) {
 
       if (AST_COMPARE) {
         const ast = parse(source, mergedOptions);
-        const astMassaged = massageAST(ast);
+        const normalizedOptions = normalizeOptions(mergedOptions);
+        const astMassaged = massageAST(ast, normalizedOptions);
         let ppastMassaged;
         let pperr = null;
         try {
           const ppast = parse(
             prettyprint(source, path, mergedOptions),
             mergedOptions
           );
-          ppastMassaged = massageAST(ppast);
+          ppastMassaged = massageAST(ppast, normalizedOptions);
```

Comment:
so this is definitely a value-add, but it's not an ideal API. for purposes of comparison, given a python AST, you can call

```python
ast.dump(tree, annotate_fields=False)
```

this will give you a string representation of the AST – not the most useful thing for what we're formatting, but you can then just do string equality to compare.

e.g. for this python 2 example:

```
In [15]: print(s)

with A() as a, B() as b:
    pass


In [16]: print(rt)
with A() as a:
    with B() as b:
        pass


In [17]: print(ast.dump(ast.parse(s), annotate_fields=False))
Module([With(Call(Name('A', Load()), [], [], None, None), Name('a', Store()), [With(Call(Name('B', Load()), [], [], None, None), Name('b', Store()), [Pass()])])])

In [18]: print(ast.dump(ast.parse(rt), annotate_fields=False))
Module([With(Call(Name('A', Load()), [], [], None, None), Name('a', Store()), [With(Call(Name('B', Load()), [], [], None, None), Name('b', Store()), [Pass()])])])
```

(the two dumped AST strings above are identical)

the difference being in this case that we'd be delegating everything to the ground truth AST handler, such that if there were a bug with the AST -> JSON conversion, we would know

the sort of essential problem from a correctness perspective is that we're not really dealing with a canonical python AST here

funny enough you can't actually assert abstract equality between ASTs in python without dumping them – at not least until https://github.com/python/cpython/pull/1368 lands

Reply:
tl;dr it would be a better upstream API to have a `parseAndMassage` hook instead of a `massageAST` hook. this is because there's a built-in way to canonicalize a python AST to a string that can be used for equality comparisons, but the functionality to convert the python AST to JSON is code under test here

Reply:
Actually, `astor.strip_tree` will clean up Python ASTs such that equality checks on them in Python are possible, but I guess there's not a good way to pretty-print the diff for test failures.

Reply:
The nested with case is something we wouldn't change in Prettier for JS. e.g. we would never change:

```js
if (foo) {
  if (bar) {
  }
}
```
to

```js
if (foo && bar) {
}
```

There's a reason we massage the AST _after_ we have printed it, as the cleanup often removes properties or nodes that are critical information required to print the code.

Reply:
that was an illustrative example, showing how this process produces identical dumped ASTs even when the source is quite different in terms of concrete syntax

as it happens, for `with`, the python 2 AST doesn't distinguish at all between those cases – hence the heuristic in https://github.com/prettier/prettier-python/pull/15/files#diff-ab7972325670f1753462f9b0e6c57900R138

like literally those two things above are identical from the perspective of the parser in python 2 (though not so in python 3)